raft/src/main/java/org/apache/kafka/snapshot/SnapshotWriter.java

```diff
@@ -135,9 +210,12 @@ public void append(List<T> records) {
 
     /**
      * Freezes the snapshot by flushing all pending writes and marking it as immutable.
+     *
+     * Also adds a {@link MetadataSnapshotFooterRecord} to the end of the snapshot
      */
     public void freeze() {
         appendBatches(accumulator.drain());
+        finalizeSnapshotWithFooter();
```

Comment:
I think we need to do this before appending the batches above.

Reply:
Would we not want to drain the pipe (here accumulator) before finalizing it (by stamping the footer). I guess your point is that the footer is yet another record which we can append and then just do one final drain?
PS right now in the code the finalize method is also forcing a drain on the accumulator. We could remove that too then. 

Let me do it that way.

Reply:
When `freeze` is called it is possible that the `accumulator` has batches. When freezing the `SnapshotWriter`, it needs to:
1. Add the footer batch to the accumulator
2. Drain any pending batches in the accumulator including the control batch appended in 1.
3. Append those batches to the underlying `RawSnapshotWriter`.

Reply:
Ahh. I was missing point 3. Understood your point. Will make the change.